src/main/java/jcifs/smb/SmbTransportImpl.java

```diff
@@ -0,0 +1,1841 @@
+     * @param resp
+     * @param path
+     * @param req
+     * @throws SmbException
+     * @throws DfsReferral
+     */
+    private void checkReferral ( Response resp, String path, RequestWithPath req ) throws SmbException, DfsReferral {
+        DfsReferralData dr = null;
+        if ( !getContext().getConfig().isDfsDisabled() ) {
+            try {
+                dr = getDfsReferrals(getContext(), path, req.getServer(), req.getDomain(), 1);
+            }
+            catch ( CIFSException e ) {
+                throw new SmbException("Failed to get DFS referral", e);
+            }
+        }
+        if ( dr == null ) {
+            if ( log.isDebugEnabled() ) {
+                log.debug("Error code: 0x" + Hexdump.toHexString(resp.getErrorCode(), 8));
+            }
+            throw new SmbException(resp.getErrorCode(), null);
+        }
+
+        if ( req.getDomain() != null && getContext().getConfig().isDfsConvertToFQDN() && dr instanceof DfsReferralDataImpl ) {
+            ( (DfsReferralDataImpl) dr ).fixupDomain(req.getDomain());
+        }
+        if ( log.isDebugEnabled() ) {
+            log.debug("Got referral " + dr);
+        }
+
+        getContext().getDfs().cache(getContext(), path, dr);
+        throw new DfsReferral(dr);
+    }
+
+
+    <T extends CommonServerMessageBlockResponse> T send ( CommonServerMessageBlockRequest request, T response ) throws SmbException {
+        return send(request, response, Collections.<RequestParam> emptySet());
+    }
+
+
+    <T extends CommonServerMessageBlockResponse> T send ( CommonServerMessageBlockRequest request, T response, Set<RequestParam> params )
+            throws SmbException {
+        ensureConnected(); /* must negotiate before we can test flags2, useUnicode, etc */
+        if ( this.smb2 && ! ( request instanceof ServerMessageBlock2 ) ) {
+            throw new SmbException("Not an SMB2 request " + request.getClass().getName());
+        }
+        else if ( !this.smb2 && ! ( request instanceof ServerMessageBlock ) ) {
+            throw new SmbException("Not an SMB1 request");
+        }
+
+        this.negotiated.setupRequest(request);
+
+        if ( response != null ) {
+            request.setResponse(response); /* needed by sign */
+            response.setDigest(request.getDigest());
+        }
+
+        try {
+            if ( log.isTraceEnabled() ) {
+                log.trace("Sending " + request);
+            }
+            if ( request.isCancel() ) {
+                doSend0(request);
+                return null;
+            }
+            else if ( request instanceof SmbComTransaction ) {
+                response = sendComTransaction(request, response, params);
+            }
+            else {
+                if ( response != null ) {
+                    response.setCommand(request.getCommand());
+                }
+                response = sendrecv(request, response, params);
+            }
+        }
+        catch ( SmbException se ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)